text/0000-execution-environments.md

```diff
@@ -0,0 +1,245 @@
+
+### Project Descriptor - `project.toml` (App Developers)
+
+An app developer may have execution environment configuration like only using a metrics agent in production or headless user agent needs in test. In order to facilitate these needs, the project descriptor will be extended by adding `exec-env` to the following tables:
+
+* `[[io.buildpacks.group]]`
+* `[[io.buildpacks.pre.group]]`
+* `[[io.buildpacks.post.group.env]]`
+* `[[io.buildpacks.build.env]]`
+
+An example would look like this:
+
+```TOML
+[_]
+schema-version = "0.3"
+
+[[io.buildpacks.group]]
+id = "buildpacks/ruby"
+version = "latest"
+
+[[io.buildpacks.group]]
+id = "buildpacks/nodejs"
+version = "latest"
+exec-env = ["production", "test"]
+
+[[io.buildpacks.group]]
+id = "buildpacks/metrics-agent"
+version = "latest"
+exec-env = ["production"]
+
+[[io.buildpacks.group]]
+id = "buildpacks/headless-chrome"
+version = "latest"
+exec-env = ["test"]
+
+[[io.buildpacks.post.group]]
+id = "buildpacks/procfile"
+version = "latest"
+
+[[io.buildpacks.build.env]]
+name = "RAILS_ENV"
+value = "production"
+exec-env = ["production"]
+
+[[io.buildpacks.build.env]]
+name = "RAILS_ENV"
+value = "test"
+exec-env = ["test"]
+
+[[io.buildpacks.build.env]]
+name = "PARALLEL_WORKERS"
+value = "4"
+exec-env = ["production"]
+```
+
+### `builder.toml` (Builder Authors)
+
+The the only table `exec-env` will be added to is `[[order.group]]` and `[[build.env]]`.
+
+### `buildpack.toml` (Buildpack Authors)
+
+The only table that `exec-env` will be added to is `[[buildpack.order.group]]`. This only is applicable for composite buildpacks.
+
+### `launch.toml` (Buildpack Authors)
+
+Not all process types make sense for every execution environment. In order to help hint to the platform the intention how a process should be used, the `exec-env` key will be added to the `[[processes]]` table.
+
+### `metadata.toml` (Platform Operators)
+
+On the platform side, the `exec-env` key will be added to `metadata.toml` in the `[[processes]]` table to mirror `launch.toml`. This will help platforms to make decisions on which processes are desired for each execution environment. `lifecycle` will list all processes and will not exclude any based on the execution environment Instead, a platform can use the available information to make a decision.
+
+## `CNB_EXEC_ENV` Environment Variable
+
+The spec will reserve the following values to help standardize execution environments:
+
+* production
+* test
+* development
+
+### Buildpack API
+
+A buildpack author will be able to determine the execution environment their buildpack is expected to build for by reading the `CNB_EXEC_ENV` environment variable. If this value is not set, a Buildpack Author can assume it's set to `production`. This will be provided for both `bin/detect` and `bin/build`.
```

Comment:
Instead of defaulting to `production`, should we have a concept of `all`? Maybe this is similar to the `exec-env` having `*`?

Reply:
@jkutner do we need a concept of "all" for an individual lifecycle execution?

Reply:
Talking with @jabrown85 , I think compiled languages like Go might want to remove the source tree from `/workspace` for production but not test which would extra complicate things.